torchvision/prototype/datasets/_builtin/sintel.py

```diff
@@ -0,0 +1,167 @@
+        next(b, None)
+        return zip(a, b)
+
+
+class IntCategoryGrouper(IterDataPipe[Tuple[Tuple[str, T], Tuple[str, T]]]):
+    def __init__(self, datapipe: IterDataPipe[Tuple[str, T]]) -> None:
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[Tuple[str, Any], Tuple[str, Any]]]:
+        for item1, item2 in pairwise(sorted(self.datapipe)):
+            if pathlib.Path(item1[0]).parent != pathlib.Path(item2[0]).parent:
+                continue
+
+            yield item1, item2
+
+
+class SINTEL(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "sintel",
+            type=DatasetType.IMAGE,
+            homepage="",
+            valid_options=dict(
+                split=("train", "test"),
+                pass_=("clean", "final"),
+            ),
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        archive = HttpResource(
+            "http://sintel.cs.washington.edu/MPI-Sintel-complete.zip",
+            sha256="",
+        )
+        return [archive]
+
+    def _classify_train_test(self, data: Tuple[str, Any], *, config: DatasetConfig) -> bool:
+        path = pathlib.Path(data[0])
+        return config.split in str(path.parent)
+
+    def _classify_archive(self, data: Tuple[str, Any], *, config: DatasetConfig) -> Union[int, None]:
+        path = pathlib.Path(data[0])
+        if config.pass_ == path.parent.parent.name and path.suffix == ".png":
+            return 0
+        elif path.suffix == ".flo":
+            return 1
+        else:
+            return None
+
+    def _read_flo(self, file: IterDataPipe[Tuple[Any, io.IOBase]]) -> Any:
+        magic = file.read(4)
+        if magic != b"PIEH":
+            raise ValueError("Magic number incorrect. Invalid .flo file")
+        w = int.from_bytes(file.read(4), "little")
+        h = int.from_bytes(file.read(4), "little")
+        data = file.read(2 * w * h * 4)
+        data = np.frombuffer(data, dtype=np.float32)
+        return data.reshape(h, w, 2).transpose(2, 0, 1)
+
+    def _flows_key(self, data: Tuple[str, Any]) -> Tuple[str, int]:
+        path = pathlib.Path(data[0])
+        category = path.parent.name
+        idx = int(FILE_NAME_PATTERN.match(path.name).group("idx"))  # type: ignore[union-attr]
+        return category, idx
+
+    def _images_key(self, data: Tuple[Tuple[str, Any], Tuple[str, Any]]) -> Tuple[str, int]:
+        return self._flows_key(data[0])
+
+    def _collate_and_decode_sample(
+        self,
+        data: Tuple[Tuple[str, io.IOBase], Any],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+        config: DatasetConfig,
+    ) -> Dict[str, Any]:
+        if config.split == "train":
+            flo, images = data
+            img1, img2 = images
+            flow_arr = self._read_flo(flo[1])
+            del images, flo
+        else:
+            # When split is `test`
+            img1, img2 = data
```

Comment:
Having different input types / sizes based on the split is an anti-pattern. Instead we should have the same structure for both splits and substitute `None` for the flo file handle in the test split. That way we can simply do

```python
flow = self._read_flo(flo_buffer) if flow_buffer else None
```